test/Identity.Test/IdentityServer/UserDecryptionOptionsBuilderTests.cs

```diff
@@ -0,0 +1,185 @@
+using Bit.Core.Auth.Enums;
+using Bit.Core.Auth.Models.Data;
+using Bit.Core.Context;
+using Bit.Core.Entities;
+using Bit.Core.Repositories;
+using Bit.Core.Services;
+using Bit.Identity.IdentityServer;
+using Bit.Identity.Utilities;
+using Bit.Test.Common.AutoFixture.Attributes;
+using NSubstitute;
+using Xunit;
+
+namespace Bit.Identity.Test.IdentityServer;
+
+public class UserDecryptionOptionsBuilderTests
+{
+    private readonly ICurrentContext _currentContext;
+    private readonly IFeatureService _featureService;
+    private readonly IDeviceRepository _deviceRepository;
+    private readonly IOrganizationUserRepository _organizationUserRepository;
+    private readonly UserDecryptionOptionsBuilder _builder;
+
+    public UserDecryptionOptionsBuilderTests()
+    {
+        _currentContext = Substitute.For<ICurrentContext>();
+        _featureService = Substitute.For<IFeatureService>();
+        _deviceRepository = Substitute.For<IDeviceRepository>();
+        _organizationUserRepository = Substitute.For<IOrganizationUserRepository>();
+        _builder = new UserDecryptionOptionsBuilder(_currentContext, _featureService, _deviceRepository, _organizationUserRepository);
+    }
+
+    [Theory, BitAutoData]
+    public async Task ForUser_WhenUserHasMasterPassword_ShouldReturnMasterPasswordOption(User user)
+    {
+        user.MasterPassword = "password";
+
+        var result = await _builder.ForUser(user).BuildAsync();
+
+        Assert.True(result.HasMasterPassword);
+    }
+
+    [Theory, BitAutoData]
+    public async Task ForUser_WhenUserDoesNotHaveMasterPassword_ShouldNotReturnMasterPasswordOption(User user)
+    {
+        user.MasterPassword = null;
+
+        var result = await _builder.ForUser(user).BuildAsync();
+
+        Assert.False(result.HasMasterPassword);
+    }
+
+    [Theory, BitAutoData]
+    public async Task WithWebAuthnLoginCredential_WhenCredentialDoesNotContainPrfKeys_ShouldNotReturnPrfOption(WebAuthnCredential credential)
+    {
+        credential.EncryptedPrivateKey = null;
+        credential.EncryptedPublicKey = null;
+        credential.EncryptedUserKey = null;
+
+        var result = await _builder.WithWebAuthnLoginCredential(credential).BuildAsync();
+
+        Assert.Null(result.WebAuthnPrfOption);
+    }
+
+    [Theory, BitAutoData]
+    public async Task WithWebAuthnLoginCredential_WhenCredentialContainsPrfKeys_ShouldReturnPrfOption(WebAuthnCredential credential)
+    {
+        credential.EncryptedPrivateKey = "encryptedPrivateKey";
+        credential.EncryptedPublicKey = "encryptedPublicKey";
+        credential.EncryptedUserKey = "encryptedUserKey";
+
+        var result = await _builder.WithWebAuthnLoginCredential(credential).BuildAsync();
+
+        Assert.NotNull(result.WebAuthnPrfOption);
+        Assert.Equal(credential.EncryptedPrivateKey, result.WebAuthnPrfOption!.EncryptedPrivateKey);
+        Assert.Equal(credential.EncryptedUserKey, result.WebAuthnPrfOption!.EncryptedUserKey);
+    }
```

Comment:
Question: would it be a good idea to consolidate these into a combined test which tested all combinations for better coverage? I tested this and it worked. 

```C#
[Theory]
    [BitAutoData(true, true, true)]    // All keys are non-null
    [BitAutoData(false, false, false)] // All keys are null
    [BitAutoData(false, false, true)]  // EncryptedUserKey is non-null, others are null
    [BitAutoData(false, true, false)]  // EncryptedPublicKey is non-null, others are null
    [BitAutoData(true, false, false)]  // EncryptedPrivateKey is non-null, others are null
    [BitAutoData(true, false, true)]   // EncryptedPrivateKey and EncryptedUserKey are non-null, EncryptedPublicKey is null
    [BitAutoData(true, true, false)]   // EncryptedPrivateKey and EncryptedPublicKey are non-null, EncryptedUserKey is null
    [BitAutoData(false, true, true)]   // EncryptedPublicKey and EncryptedUserKey are non-null, EncryptedPrivateKey is null
    public async Task WithWebAuthnLoginCredential_VariousKeyCombinations_ShouldReturnCorrectPrfOption(
        bool hasEncryptedPrivateKey,
        bool hasEncryptedPublicKey,
        bool hasEncryptedUserKey,
        WebAuthnCredential credential)
    {
        credential.EncryptedPrivateKey = hasEncryptedPrivateKey ? "encryptedPrivateKey" : null;
        credential.EncryptedPublicKey = hasEncryptedPublicKey ? "encryptedPublicKey" : null;
        credential.EncryptedUserKey = hasEncryptedUserKey ? "encryptedUserKey" : null;

        var result = await _builder.WithWebAuthnLoginCredential(credential).BuildAsync();

        if (credential.GetPrfStatus() == WebAuthnPrfStatus.Enabled)
        {
            Assert.NotNull(result.WebAuthnPrfOption);
            Assert.Equal(credential.EncryptedPrivateKey, result.WebAuthnPrfOption!.EncryptedPrivateKey);
            Assert.Equal(credential.EncryptedUserKey, result.WebAuthnPrfOption!.EncryptedUserKey);
        }
        else
        {
            Assert.Null(result.WebAuthnPrfOption);
        }
    }
```

Reply:
Nice, thanks, copy-pasted! Added you as co-author on the commit since you basically wrote the whole thing 😁 

![image](https://github.com/bitwarden/server/assets/2285588/daf2c156-9a5f-41c4-ab00-7c24e6f305fb)